core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/protocol/restxml/XmlProtocolMarshaller.java

```diff
@@ -0,0 +1,179 @@
+                                                     .request(request)
+                                                     .build();
+    }
+
+    private Request<OrigRequestT> fillBasicRequestParams(OperationInfo operationInfo, OrigRequestT originalRequest) {
+        Request<OrigRequestT> request = new DefaultRequest<>(originalRequest, operationInfo.serviceName());
+        request.setHttpMethod(operationInfo.httpMethodName());
+        request.setResourcePath(UriResourcePathUtils.addStaticQueryParametersToRequest(request, operationInfo.requestUri()));
+        return request;
+    }
+
+    @Override
+    public Request<OrigRequestT> marshall(SdkPojo pojo) {
+        doMarshall(pojo);
+        return finishMarshalling();
+    }
+
+    public void doMarshall(SdkPojo pojo) {
+        for (SdkField<?> field : pojo.sdkFields()) {
+            Object val = resolveValue(field.get(pojo), field);
+
+            if (isBinary(field, val)) {
+                request.setContent(((SdkBytes) val).asInputStream());
+                setContentTypeHeaderIfNeeded("binary/octet-stream");
+
+            } else if (isExplicitPayloadMember(field) && val instanceof String) {
+                if (val != null) {
+                    byte[] content = ((String) val).getBytes(StandardCharsets.UTF_8);
+                    request.setContent(new ByteArrayInputStream(content));
+                    request.addHeader(CONTENT_LENGTH, Integer.toString(content.length));
+                    if (!request.getHeaders().containsKey(CONTENT_MD5)) {
+                        request.addHeader(CONTENT_MD5, Md5Utils.md5AsBase64(content));
+                    }
+                }
+            } else {
+                MARSHALLER_REGISTRY.getMarshaller(field.location(), field.marshallingType(), val)
+                                   .marshall(val, marshallerContext, field.locationName(), (SdkField<Object>) field);
+            }
+        }
+    }
+
+    private Request<OrigRequestT> finishMarshalling() {
+        // Content may already be set if the payload is binary data.
+        if (request.getContent() == null && marshallerContext.xmlGenerator() != null) {
+            String content = marshallerContext.xmlGenerator().stringWriter().getBuffer().toString();
+
+            if (!content.isEmpty()) {
+                request.setContent(new StringInputStream(content));
+                request.addHeader("Content-Length", Integer.toString(content.getBytes(StandardCharsets.UTF_8).length));
+                setContentTypeHeaderIfNeeded("application/xml");
+            }
+        }
+
+        return request;
+    }
+
+    private boolean isBinary(SdkField<?> field, Object val) {
+        return isExplicitPayloadMember(field) && val instanceof SdkBytes;
+    }
+
+    private boolean isExplicitPayloadMember(SdkField<?> field) {
+        return field.containsTrait(PayloadTrait.class);
+    }
+
+    private Object resolveValue(Object val, SdkField<?> marshallingInfo) {
```

Comment:
I wonder if this is something we should move into the SdkField class since it's common across all protocols.